### PR TITLE
Reverse sodium_hex2bin and try to fix hex2bin

### LINKED ISSUE #286: error: hex2bin more strictly aligned
Output of `cargo clippy`:

```
error: casting from `*mut u8` to a more-strictly-aligned pointer (`*mut *const i8`)
   --> src/utils.rs:113:13
    |
113 |             hex_end as *mut _,
    |             ^^^^^^^^^^^^^^^^^
    |
    = note: #[deny(clippy::cast_ptr_alignment)] on by default
    = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.212/index.html#cast_ptr_alignment

error: aborting due to previous error
```

https://rust-lang.github.io/rust-clippy/master/index.html#cast_ptr_alignment

> Dereferencing the resulting pointer may be **undefined behavior**.

@lovesh need your help

### PR DESCRIPTION
**Pull request by kpp**
It passes `cargo test` and `cargo clippy` (at least no segfault from clippy). But I wonder whether these bindings are idiomatic or not.

Closes #286

### CODE REVIEW

File: src/utils.rs [hunk 1]
```diff
@@ -87,34 +87,42 @@ pub fn bin2hex(hex: &mut [u8], bin: &[u8]) -> Result<(), ()> {
...
-/// string couldn't be fully parsed, but a valid pointer for `hex_end` was not provided
+/// `hex2bin()` will return Err<()> if `bin` is not sufficient to store the parsed string
 /// Refer docs at https://download.libsodium.org/doc/helpers#hexadecimal-encoding-decoding
 pub fn hex2bin(
     bin: &mut [u8],
     hex: &[u8],
     ignore: Option<&[u8]>,
     bin_len: &mut usize,
-    hex_end: Option<&mut [u8]>,
+    hex_end: Option<&mut isize>,
```
Comment (dnaq) [inline src/utils.rs]: Not good. hex_end should be a pointer to a pointer, so it’s wrong here.

Also exposing that is completely unidiomatic in a rust api, and idiomatic way would be to return ‘Result<&[u8]> where the return value is a sub slice of ‘hex’. But to be honest that’s a very C-ish way to write an api, and I don’t see the need for exposing ‘hex_end’ at all.

File: src/utils.rs [hunk 2]
```diff
@@ -87,34 +87,42 @@ pub fn bin2hex(hex: &mut [u8], bin: &[u8]) -> Result<(), ()> {
...
 ) -> Result<(), ()> {
-    let ignore = ignore.map_or(null(), |p| p.as_ptr());
-    let hex_end = hex_end.map_or(null_mut(), |p| p.as_mut_ptr());
+    let ignore = ignore.map_or(ptr::null(), |p| p.as_ptr());
 
-    unsafe {
-        let r = ffi::sodium_hex2bin(
+    let mut hex_end_ptr: *const i8 = ptr::null();
```
Comment (dnaq) [inline src/utils.rs]: Sorry. Misstated above.

File: src/utils.rs (hunk 1 above)
Comment (kpp) [inline src/utils.rs]: @dnaq I don't like these `Option<&...>`. Looks awful

Comment (dnaq) [reply src/utils.rs]: Agreed

**Comment by dnaq** [comment]
Please just remove all of it, we can have a new discussion about the respective merits and demerits of this functionality at a later stage. For now it just has to disappear.

**Comment by kpp** [comment]
> Please just remove all of it

Done

**Pull request merged**